### PR TITLE
Add RFDiffusionModel TorchModel wrapper with self-conditioning

### PR DESCRIPTION
**Pull request by chidwipak**
## Description

Fix #4747

Adds `RFDiffusionModel`, a DeepChem `TorchModel` wrapper around the protein backbone diffusion layers. This PR adds the training generator, diffusion loss, coordinate normalization, checkpoint handling for normalization statistics, optional self-conditioning, and `generate()` sampling.

The model is intentionally documented as a baseline DDPM scaffold for protein backbone coordinates. It does not yet implement the full SE(3)-equivariant RFDiffusion architecture; the rigid-frame/IPA work is planned separately.

Implementation details:

- wraps `BackboneDiffusion` in a DeepChem `TorchModel`
- supports `model.fit(dataset)` on object-array protein backbone datasets
- samples random timesteps and trains with a noise-prediction MSE objective
- adds residue masks for variable-length padded batches
- respects dataset/sample weights, including zero-weight padded repeats from `pad_batches=True`
- raises clearly for unsupported predict/uncertainty generator modes
- validates generation arguments and sequence lengths
- preserves train/eval state during generation
- saves/restores denormalization statistics with checkpoints
- adds tests for training, variable lengths, checkpoint restore, self-conditioning, validation errors, and sample-weight handling

Validation run locally:

```bash
/home/chidwipak/Gsoc2026/deepchem/venv/bin/python -m pytest deepchem/models/torch_models/tests/test_rfdiffusion_layers.py deepchem/models/torch_models/tests/test_rfdiffusion_model.py -q
# 55 passed, 9 warnings
```

The warnings are PyTorch `torch.load(weights_only=False)` future warnings from checkpoint restore paths.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by chidwipak** [comment]
Closing this PR to split the architecture into smaller, more modular PRs for easier review.
